modules/core/include/opencv2/core/cvdef.h

```diff
@@ -744,89 +744,42 @@ __CV_ENUM_FLAGS_BITWISE_XOR_EQ   (EnumType, EnumType)
 #endif
 
 
-/****************************************************************************************\
-*                      CV_NODISCARD attribute (deprecated, GCC only)                     *
-* DONT USE: use instead the standard CV_NODISCARD_STD macro above                        *
-*           this legacy method silently fails to issue warning until some version        *
-*           after gcc 6.3.0. Yet with gcc 7+ you can use the above standard method       *
-*           which makes this method useless. Don't use it.                               *
-* @deprecated use instead CV_NODISCARD_STD                                               *
-\****************************************************************************************/
-#ifndef CV_NODISCARD
-#  if defined(__GNUC__)
-#    define CV_NODISCARD __attribute__((__warn_unused_result__))
-#  elif defined(__clang__) && defined(__has_attribute)
-#    if __has_attribute(__warn_unused_result__)
-#      define CV_NODISCARD __attribute__((__warn_unused_result__))
-#    endif
-#  endif
-#endif
-#ifndef CV_NODISCARD
-#  define CV_NODISCARD /* nothing by default */
-#endif
-
-
 /****************************************************************************************\
 *                                    C++ 11                                              *
 \****************************************************************************************/
-#ifndef CV_CXX11
-#  if __cplusplus >= 201103L || (defined(_MSC_VER) && _MSC_VER >= 1800)
-#    define CV_CXX11 1
-#  endif
-#else
-#  if CV_CXX11 == 0
-#    undef CV_CXX11
+#ifdef __cplusplus
+// MSVC was stuck at __cplusplus == 199711L for a long time, even where it supports C++11,
+// so check _MSVC_LANG instead. See:
+// <https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus>
+#  if defined(_MSC_VER) && (_MSC_VER < 1800)
+#    error "OpenCV 4.x+ requires enabled C++11 support"
+#  elif __cplusplus < 201103L
```

Comment:
It still does not work, first condition passes, but second condition still fails.
```
C:\build\precommit_windows64\4.x\opencv\modules\core\include\opencv2/core/cvdef.h(757): fatal error C1189: #error:  "OpenCV 4.x+ requires enabled C++11 support" 
```
Should be something like this:
```.cpp
#if GOOD_MSVC_VERSION
 // ok
#elif BAD_MSVC_VERSION || BAD_CPLUSPLUS_VERSION
 // error
#endif
```
How can it be optimized :thinking: ?

Reply:
Maybe:
```.cpp
#if MSVC
  #if BAD_VERSION
    // error
  #endif
#elif BAD_CPLUSPLUS
  // error
#endif
```

Reply:
Yes, that makes sense; done. Sorry this is taking a few iterations, like I said, I don't have MSVC locally. :)